app/models/invite.rb

```diff
@@ -160,6 +182,39 @@ def self.create_invite_by_email(email, invited_by, opts = nil)
     invite
   end
 
+  def self.generate_multiple_use_invite_link(invited_by:, max_redemptions_allowed: 5, expires_at: 1.month.from_now, group_ids: nil)
+    max_redemptions_allowed = max_redemptions_allowed.to_i
+    unless max_redemptions_allowed.between?(2, SiteSetting.invite_link_max_redemptions_limit)
```

Comment:
I don't think this is the right place to do this check as it assumes all multiple use invite link has to go through this code path. Instead, I think this should be an active record validation.

Reply:
Done.

Reply:
Actually I had to revert this change because single use invite has `max_redemptions_allowed` default to `1` which will fail active record validation. I added another active record validation which checks for numericality and verifies that value is greater than 0.

Reply:
We should only run the validation for multiple invite links so that it doesn't affect single use invites.

Reply:
To identify if an invite is multiple invite link we check for `max_redemptions_allowed > 1`... so how can we add condition of the same field that we are validating?

Reply:
I thought we were leaving `max_redemptions_allowed` blank for single invite so that the database default would kick in?

Reply:
We have `add_column :invites, :max_redemptions_allowed, :integer, null: false, default: 1` so single invite should just leave `max_redemptions_allowed` blank

Reply:
Does this validation looks good?

```
validates :max_redemptions_allowed,
    allow_blank: true,
    numericality: { only_integer: true, greater_than_or_equal_to: 2, less_than_or_equal_to: SiteSetting.invite_link_max_redemptions_limit, allow_blank: true }
```

I am having trouble with above validation clause when creating single use invite link. It's failing with `Validation failed: Max redemptions allowed must be greater than or equal to 2` even when `max_redemptions_allowed` is left blank.

Reply:
I’ll recommend writing a custom validation instead of trying to work with the Rails options here. Much easier to control what we want to do. You can add a custom validation with validate :some_method

Reply:
I tried adding custom validation but when I check for `max_redemptions_allowed.present?` it returns true even for single use invites. It seems like `max_redemptions_allowed` defaults to 1 at the time of custom validation, even when I have not set it explicitly.

Reply:
O I think there is some active record magic here. Can you try setting `max_redemptions_allowed: nil` for single invites?

Reply:
In that case I'm getting `ActiveRecord::NotNullViolation`. I am really inclined to keep this check in `generate_multiple_use_invite_link` since we're not going to create multi use invites anywhere else anytime soon. Also, we are already checking for numericality and min value of `1` in active record validation.

Reply:
> I am really inclined to keep this check in `generate_multiple_use_invite_link` since we're not going to create multi use invites anywhere else anytime soon.

The rational for adding this as an ActiveRecord validation is because we don't know who/when will come along and create an invite link in the future without going through the proper methods. By having it as a ActiveRecord validation, we get to ensure consistency before the record hits the database. Can you see if the following works? 

```
  validate :ensure_max_redemptions_allowed

  def ensure_max_redemptions_allowed
    if self.max_redemptions_allowed.nil?
      self.max_redemptions_allowed = 1
    else
      # validation
    end
  end
```

Reply:
The validation here is really to ensure that if someone provides us with a value of more than `1`,  we need to ensure that it falls between `2` and `SiteSetting.max`

Reply:
Actually, can't we use the lack of an `email` to differentiate between a multi invite link vs a single invite link? Or can there be a single invite link? If so, then we just have to validate that the value falls between 1 and SiteSetting max.

Reply:
We can't check for `email` because we still have official `discourse-invite-tokens` plugin which allows maximum redemption of `1` but still has no email attached to it.

I'll try your validation which checks for range `1..SiteSetting max`.

Reply:
Done.